examples/audio_server/server.cc

```diff
@@ -0,0 +1,132 @@
+#include <memory>
+#include <mutex>
+
+#include "../../blocks/http/api.h"
+#include "../../bricks/dflags/dflags.h"
+#include "../../bricks/sync/waitable_atomic.h"
+
+DEFINE_uint16(port, 8080, "The local port to use.");
+DEFINE_uint32(n, 3, "Max number of audio streams");
+
+const std::string cmd_kill = "kill";
+
+struct SharedState final {
+  uint32_t active_channels = 0;
+  std::map<std::string, std::string> channel_control;
+  bool channel_exists(std::string channel_id) const{
+    return channel_control.find(channel_id) != channel_control.end();
+  }
+};
+
+struct ControlSignal final {
+  std::string apply= "";
+  bool stop = false;
+};
+
+int main(int argc, char** argv) {
+  ParseDFlags(&argc, &argv);
+
+  try {
+    auto& http = HTTP(current::net::AcquireLocalPort(FLAGS_port));
+
+    current::WaitableAtomic<SharedState> safe_state;
+    std::map<std::string, std::thread> threads;
+
+    // Need to join and remove finished sessions
+    auto joiner = std::thread([&threads, &safe_state]{
+      while(true) {
+        std::vector<std::string> to_join;
+        for (auto& pair : threads) {
+          std::string channel_id = pair.first;
+          auto has_channel = safe_state.MutableUse([channel_id](SharedState& state) {
+            return state.channel_exists(channel_id);
+          });
+          if (!has_channel) {
+            to_join.push_back(pair.first);
+            pair.second.join();
+          }
+        }
+        for(auto& to_drop: to_join){
+          threads.erase(to_drop);
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+      }
+    });
+
+    auto scope = http.Register("/", [&safe_state, &threads](Request r) {
+      std::string channel_id = r.body;
+
+      auto has_channel = safe_state.MutableUse([channel_id](SharedState& state) { return state.channel_exists(channel_id); });
+      if(has_channel){
+        r("error: channel already exists\n");
+        return;
+      }
+      uint32_t n_chans = safe_state.MutableUse([](SharedState& state) { return state.active_channels; });
+      if(n_chans == FLAGS_n){
+        r("error: too many channels\n");
+        return;
+      }
+
+      // Create new channel
+      safe_state.MutableUse([channel_id](SharedState& state) {
+        state.channel_control[channel_id] = "";
+        state.active_channels++;
+      });
+      threads[channel_id] = std::thread([&safe_state, channel_id]() {
```

Comment:
And this code, I think, relies on the fact that the HTTP handler for a particular port is single-threaded. Which is correct, but why not put the threads into the `WaitableAtomic<>` state too?

Reply:
I.e. it is "correct" so that `cout` output will not get garbled, but overall we can do better =)

Reply:
done, moved threads to WaitableAtomic